wallet/wallet.go

```diff
@@ -683,108 +676,136 @@ func (w *Wallet) recovery(startHeight int32) error {
 	if err != nil {
 		return err
 	}
-	tx, err := w.db.BeginReadWriteTx()
-	if err != nil {
-		return err
-	}
-	txMgrNS := tx.ReadBucket(wtxmgrNamespaceKey)
-	credits, err := w.TxStore.UnspentOutputs(txMgrNS)
-	if err != nil {
-		tx.Rollback()
-		return err
-	}
-	addrMgrNS := tx.ReadWriteBucket(waddrmgrNamespaceKey)
-	err = recoveryMgr.Resurrect(addrMgrNS, scopedMgrs, credits)
+	err = walletdb.View(w.db, func(tx walletdb.ReadTx) error {
+		txMgrNS := tx.ReadBucket(wtxmgrNamespaceKey)
+		credits, err := w.TxStore.UnspentOutputs(txMgrNS)
+		if err != nil {
+			return err
+		}
+		addrMgrNS := tx.ReadBucket(waddrmgrNamespaceKey)
+		return recoveryMgr.Resurrect(addrMgrNS, scopedMgrs, credits)
+	})
 	if err != nil {
-		tx.Rollback()
 		return err
 	}
 
-	// We'll also retrieve our chain backend client in order to filter the
-	// blocks as we go.
-	chainClient, err := w.requireChainClient()
+	// We'll then need to determine the range of our recovery. This properly
+	// handles the case where we resume a previous recovery attempt after a
+	// restart.
+	startHeight, bestHeight, err := w.syncHeightRange(
+		chainClient, birthdayBlock,
+	)
 	if err != nil {
-		tx.Rollback()
 		return err
 	}
 
-	// We'll begin scanning the chain from the specified starting height.
-	// Since we assume that the lowest height we start with will at least be
-	// that of our birthday, we can just add every block we process from
-	// this point forward to the recovery batch.
-	err = w.scanChain(startHeight, func(height int32,
-		hash *chainhash.Hash, header *wire.BlockHeader) error {
-
-		recoveryMgr.AddToBlockBatch(hash, height, header.Timestamp)
+	// Now we can begin scanning the chain from the specified starting
+	// height. Since the recovery process itself acts as rescan, we'll also
+	// update our wallet's synced state along the way to reflect the blocks
+	// we process and prevent rescanning them later on.
+	//
+	// NOTE: We purposefully don't update our best height since we assume
+	// that a wallet rescan will be performed from the wallet's tip, which
+	// will be of bestHeight after completing the recovery process.
+	var (
+		height = startHeight
+		blocks []*waddrmgr.BlockStamp
+	)
 
-		// We'll checkpoint our current batch every 2K blocks, so we'll
-		// need to start a new database transaction. If our current
-		// batch is empty, then this will act as a NOP.
-		if height%recoveryBatchSize == 0 {
-			blockBatch := recoveryMgr.BlockBatch()
-			err := w.recoverDefaultScopes(
-				chainClient, tx, addrMgrNS, blockBatch,
-				recoveryMgr.State(),
-			)
+	for height < bestHeight {
```

Comment:
Any specific reason not to keep the existing single loop with `if height%recoveryBatchSize == 0` check?

Reply:
This logic had to be changed in order to handle the recovery syncing blocks that are not part of our birthday but required as part of the wallet maintaining the last 10K block hashes. This check still exists within the inner loop as `for len(recoveryMgr.BlockBatch()) < recoveryBatchSize`.

Reply:
So how is this different from 
```go
for height < bestHeight {
     if height >= birthday.height {
          addToBatch()
     }

     if len(blockBatch) >= batchSize || height == bestBlock {
            recoverScopes()
            resetBlockBatch()
     }
```

Not that it matters that much, but I find a single loop easier to reason about then two, and I'm trying to understand why two is needed.

Reply:
Done. Much cleaner now :)